nimbus/src/enrollment.rs

```diff
@@ -179,55 +525,117 @@ pub fn reset_enrollment(
+            // XXX - do we want specific errors for this kind of thing?
+            log::warn!("No such experiment '{}'", experiment_slug);
+            return Ok(vec![]);
+        }
+        Some(e) => e,
+    };
+    if !exp
+        .branches
+        .iter()
+        .map(|b| &b.slug)
+        .any(|branch_slug| branch == branch_slug)
+    {
+        // XXX - do we want specific errors for this kind of thing?
+        log::warn!(
+            "No such branch '{}' for experiment {}",
+            branch,
+            experiment_slug
+        );
+        return Ok(vec![]);
+    }
     let enrollment = ExperimentEnrollment {
         slug: experiment_slug.to_string(),
-        status: EnrollmentStatus::Enrolled {
-            reason: EnrolledReason::OptIn,
-            branch: branch.to_string(),
-        },
+        status: EnrollmentStatus::enrolled(EnrolledReason::OptIn, branch),
+    };
+    let enrollment_id = if let EnrollmentStatus::Enrolled { enrollment_id, .. } = &enrollment.status
+    {
+        enrollment_id
+    } else {
+        unreachable!(); // This is really dumb.
     };
-    let mut writer = db.write()?;
     db.get_store(StoreId::Enrollments)
         .put(&mut writer, experiment_slug, &enrollment)?;
     writer.commit()?;
-    Ok(())
+    Ok(vec![EnrollmentChangeEvent::new(
+        &enrollment.slug,
+        enrollment_id,
+        branch,
+        None,
+        EnrollmentChangeEventType::Enrollment,
+    )])
 }
 
-pub fn opt_out(db: &Database, experiment_slug: &str) -> Result<()> {
-    // As above - check experiment exists?
+pub fn opt_out(db: &Database, experiment_slug: &str) -> Result<Vec<EnrollmentChangeEvent>> {
+    let mut writer = db.write()?;
+    let exp_store = db.get_store(StoreId::Experiments);
+    let _exp = match exp_store.get::<Experiment>(&writer, experiment_slug)? {
+        None => {
+            // XXX - do we want specific errors for this kind of thing?
+            log::warn!("No such experiment '{}'", experiment_slug);
+            return Ok(vec![]);
+        }
+        Some(e) => e,
+    };
     let enrollment = ExperimentEnrollment {
         slug: experiment_slug.to_string(),
-        status: EnrollmentStatus::OptedOut,
+        status: EnrollmentStatus::Disqualified {
+            reason: DisqualifiedReason::OptOut,
+            // TODO: We should use the branch we're enrolled in if we had an enrollment
+            // But what do we do if we did not? I assume we want to support the "opt out
+            // before enrollment ever happens use case". Should we just make "branch" optional?
+            branch: "not-specified".to_owned(),
+            enrollment_id: "todo".to_owned(),
+        },
     };
-    let mut writer = db.write()?;
     db.get_store(StoreId::Enrollments)
         .put(&mut writer, experiment_slug, &enrollment)?;
     writer.commit()?;
-    Ok(())
+    Ok(vec![]) // TODO: we should probably emit a "disqualified" event depending on the previous enrollment status
```

Comment:
I like the way that `evolve_enrollment` delegates to the `ExperimentEnrollment` via a kind of event-handler interface, and I wonder if leaning into that even more could help clean up these other state-changing functions. For example, rather than creating the new `ExperimentEnrollment` directly, you can imagine this calling an `existing_enrollment.on_opt_out` method that transitions to the appropriate new state, maintains any existing enrollment id, and returns a change event.

Likewise, you can imagine `on_opt_in` and `on_reset` methods that could be used by the functions above.

I think this would help us encapsulate the state-to-state transitions and make it obvious where and how to guard against illegal transitions.